ENH: refactor zero-filling and expose dtype API slot for it

ngoldbaum opened the pull request:
This refactors zero-filling to be controlled by dtype-specific logic. It also moves all zero-filling logic inside `PyArray_NewFromDescr_int` and removes the `_zerofill` helper, which now has no consumers.

I experimented with making the API slot only provide the initial zero value and let numpy fill the buffer with that value by doing e.g. repeated `memcpy` calls, but this ended up not being sufficient, since in principle each zero value might contain embedded references that must be unique, so we need control over filling the entire array buffer. It also makes it slightly simpler to handle object and void dtypes, since I can just move the existing logic from `_zerofill` to the new `object_fill_zero_value` function I added.

I didn't see the need to add a test for this to the scaled float dtype, since the legacy dtypes should use all the code paths I modified, but am happy to add such a test if that's desired.

ngoldbaum committed:
MAINT: correct typos in dtype API documentation comments

ngoldbaum committed:
ENH: Refactor special zero-filling to be managed by the DType

seberg commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -801,14 +805,29 @@ PyArray_NewFromDescr_int(
+             * fill_zero_value to zero out the array with the appropriate
+             * special zero value for the dtype. VOID arrays have
+             * fill_zero_value set but may or may not contain objects, so
+             * always allocate with calloc if zeroed is set.
```
If they contain objects, they should have `NPY_NEEDS_INIT` set?

ngoldbaum replied:
They should. The comment is referring to operations like `np.zeros(shape=10, dtype=[("A", "M8[s]")])`. It doesn't contain objects but it has `fill_zero_value` set since it's an `NPY_VOID` array, so unless I do something special for void dtypes the array buffer would get allocated with `malloc`. Is there a way I can rewrite the comment to make that clearer?

seberg replied:
Hmmmm... I think the important point is that `PyArray_FillObjectArray` which is called later assumes that everything not object is already zeroed?  So that is why you need to zero here for that case.

(That seems like somewhat weird logic, but lets look at it after the other changes, it might just go away by then anyway.)

seberg commented on numpy/core/src/multiarray/dtypemeta.c:
```diff
@@ -698,6 +698,28 @@ object_common_dtype(
+object_fill_zero_value(PyArrayObject *arr)
+{
+    PyObject *zero = PyLong_FromLong(0);
+    PyArray_FillObjectArray(arr, zero);
```
This should be `raw_array_assign_scalar` (or `PyArray_FillWithScalar`).  We should try to remove `PyArray_FillObjectArray`.

seberg commented on numpy/core/src/multiarray/dtypemeta.c:
```diff
@@ -698,6 +698,28 @@ object_common_dtype(
+void_fill_zero_value(PyArrayObject *arr)
+{
+    if (PyDataType_REFCHK(PyArray_DESCR(arr))) {
+        if (object_fill_zero_value(arr) < 0) {
```
OK, I guess here there is a reason to keep that `PyArray_FillObjectArray` thing around for this...

seberg reviewed (commented):
Thanks for looking into this, I think it has the potential to clean up a quite a bit more mess.  Unfortunately, I really would like to try to stick to that design goal of not using array objects (especially for something that is very fundamental).

ngoldbaum commented on numpy/core/include/numpy/_dtype_api.h:
```diff
@@ -199,6 +199,14 @@ typedef int (get_reduction_initial_function)(
+ *
+ * @returns -1 or 0 indicating error and arr being successfully filled.
+ */
+typedef int (fill_initial_function)(PyArrayObject *arr);
```
OK, fair enough! I think it might be possible to write this using a traversal loop. I will poke around next week.

seberg replied:
Cool, thanks.  I don't have any opinion right now about whats better.  The first is neat, but the second seems simpler and close to how identities currently work...

ngoldbaum replied:
Where in numpy is the ufunc identity buffer setup? I’ll take a look at that first since I’m not familiar with it.

seberg replied:
Its the `get_reduction_initial_function  *get_reduction_initial;` and defined here: https://github.com/numpy/numpy/blob/fd50e7f155d4c3313c71f60ab8af9bc4b2e6e354/numpy/core/include/numpy/_dtype_api.h#L178-L200

The actual implementation is more annoying, since its set up to fall back to the one on the ufunc: https://github.com/numpy/numpy/blob/1bbe5821cc6afa134221367640380f403fa8d011/numpy/core/src/umath/legacy_array_method.c#L266

(which is all a bit annoying and I wouldn't mind making it nicer, but it helps support complicated reduce initials.)

---

But of course what we would need here is much simpler, since it only needs the descr.

mhvk replied:
I agree with not having things like this on the dtype. Having a `ufunc` would seem to make a lot of sense to me. Could be an actual `zeros(out=array`, or be more like the strange private `_ones_like` that already exists (no idea why! but it is covered by astropy units...). For zero filling it would then be `zeros_like(array, out=array)`.  But another logical concept is that for dtypes that do not have strange non-identical zeros, one effectively just uses whatever `np.copyto` does.

(Sorry, I have *not* really looked at this PR, just really read the comments here... Ignore if this is all irrelevant.)

seberg replied:
I had thought about a ufunc, but overall still lean against it.

My thoughts are roughly:
* Creating a zero filled array seems very core and doing ufunc dispatching is a bit of work for something simple.  It *feels* a bit much if I think of an exposing API like `PyArray_FillBuffer(dtype, ptr, num, stride, NPY_ZEROVAL)`  (that is partially a lie, because we made sure ufuncs are now set up in a way to make exactly this fairly easy!).
* Its just seems like more hassle for the DType author to implement for something pretty basic (but of course users will implement many ufuncs anyway).

Another path would be an `ArrayMethod` (i.e. ufunc loop) that lives on the DType (just like casts are ufuncs loops/ArrayMethods that don't live on a ufunc object).

For clearing (dealloc+NULLint of array values) I had for a simplified traversal loop getter.  I don't like splitting from `ArrayMethod`, but it seemed like a lot of set up for something very basic.  I do translate this to `zero()`.

Now ufuncs seem like they should have some advantages (maybe for subclasses?), but I can't really put my finger on anything concrete right now except a certain beauty of aiming that as much as possible should a ufunc.

---

N.B.: I suspect this will continue to include "zero", "one", and also "empty" (we do need this for `object()` although I am not sure we will need it generally).
In principle one could continue down the rabbit hole with the typical `ufunc.identity` values, and `NA` filling might be a thing at some point.

mhvk replied:
Yeah, it is no more than thinking that there is a certain beauty and simplicity if a `DType` interacts with arrays only via ufuncs. Of course, regardless the `DType` has to have a way to tell what "zero" is (though there is an obvious default).

ngoldbaum committed:
MAINT: refactor zero-filling to use a traversal loop

ngoldbaum commented:
I re-implemented the API hook to use a traversal loop. Instead of using `PyArray_FillObjectArray` I now manually insert pointers to the zero `PyObject*` for object arrays and use the `_fillobject` helper used by `PyArray_FillObjectArray` for void arrays. In principle I could have avoided `_fillobject` but it seemed better to avoid duplicating the somewhat complicated recursive logic it uses.

I considered following the initial reduction API slot but decided to use a traversal loop instead both to ease implementation for me (the existing approach is basically a traversal loop, so it was straightforward to translate) and to allow more flexibility for the dtype author. I suspect that this is a niche case so in my opinion it's OK to require a little extra boilerplate if that buys dtype authors more flexibility.

seberg commented:
Thanks, I will have a look tomorrow.  

> In principle I could have avoided _fillobject but it seemed better to avoid duplicating the somewhat complicated recursive logic it uses.

We should definitely change it to use a pattern like the clear traversal loop (and hopefully remove the other place at some point).  But, we can follow up on that too.

ngoldbaum commented:
> we should not figure out the structure again on every elment, its very slow

Good point! I know next to nothing about structured dtypes so got scared of trying to heavily refactor `_fillobject`. I agree that it’s doing a lot of work that could be amortized over the whole array.

seberg commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -816,6 +840,23 @@ PyArray_NewFromDescr_int(
+        /*
+         * If the array needs special dtype-specific zero-filling logic, do that
+         */
+        if (zeroed && (fill_zero_info.func != NULL)) {
```
```suggestion
        if (NPY_UNLIKELY(zeroed && (fill_zero_info.func != NULL))) {
```
Probably doesn't matter, but we could flag it as unlikely (the effect is that the compiler will move this code to the end, so if it isn't hit it can continue quicker with what follows).

seberg commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -784,6 +784,33 @@ PyArray_NewFromDescr_int(
     if (data == NULL) {
+        NPY_traverse_info fill_zero_info;
+        /* float errors do not matter and we do not release GIL */
+        NPY_ARRAYMETHOD_FLAGS zero_flags = PyArrayMethod_MINIMAL_FLAGS;
```
```suggestion
        NPY_ARRAYMETHOD_FLAGS zero_flags;
```
No need here (not that it matters).  It's relevant for structured dtypes or chained functions only.

seberg commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -784,6 +784,33 @@ PyArray_NewFromDescr_int(
+        NPY_traverse_info fill_zero_info;
+        /* float errors do not matter and we do not release GIL */
+        NPY_ARRAYMETHOD_FLAGS zero_flags = PyArrayMethod_MINIMAL_FLAGS;
+        NPY_traverse_info_init(&fill_zero_info);
```
We need to pair this with the free/clear function, unfortunately (this will be important for structured dtypes).

One path may be to not support/check for `fill_zero_info.func == NULL` here, but only use `get_fill_zero_loop == NULL` to make the decision about calloc.  That would move this whole logic down to after the calloc choice was made.  (structured dtypes won't be able to signal calloc is good enough, but :shrug:)

Alternatively, we just need to add the cleanup either in `fail:` of before all relevant goto fails.

ngoldbaum replied:
Oh oops, I missed the existence of `NPY_traverse_info_xfree` looking at the traversal code, thank you very much for catching this.

ngoldbaum replied:
I ended up refactoring so the cleanup goes in the `fail` block. This does lead to allocating and initializing the traverse info struct in the fast path, but it's a tiny struct so I don't think that's a problem. Let me know if you'd like further massaging on this.

seberg commented on numpy/core/src/multiarray/dtypemeta.c:
```diff
@@ -524,6 +525,42 @@ void_common_instance(PyArray_Descr *descr1, PyArray_Descr *descr2)
+        *out_loop = NULL;
+    }
+    return 0;
+}
```
We should move these to `dtype_traversal.c`, also because that shows how to go about implementing a proper void version?  I had put the clear function setup in `dtype_traversal.c`.  That was the wrong choice I think, having it here is better probably.  (Because I came from the angle of making it similar to casts.)

seberg replied:
(Just as a note, these two functions really need a proper traversal implementation.)

ngoldbaum replied:
Sorry, a little confused by this comment. Should I move these functions to `dtype_traversal.c` to go with where the clear setup is right now? Or should I move the dtype-specific clear functions here? I'd slightly prefer the latter just so all the dtype-specific functionality for the builtin dtypes are in one file.

seberg replied:
Wait, I think I put the comment on the wrong line, I wanted to put this on the `dtypemeta.c` loop and get_loop implementations, sorry.  Does it make sense then?

ngoldbaum replied:
I’m confused about what you want me to do, because your comment starts with a suggestion to move the functions to the other file, but then says it was a mistake to put the clear functions there and they should probably live in this file. 

I agree with that and I think it probably makes sense to move the clear get_loop and loop implementations to this file and leave the zero-filling ones I added in this PR here.

seberg replied:
Arrrrrg!  Attaching the clear functions to the dtype is currently in `dtype_transfer.c` (grouped with the cast setups).  That is weird, and I would agree with moving it into `dtypemeta.c`.

The actual clear traversal loops and get_loops are currently in `dtype_traversal.c` and I think I prefer that because it is quite a lot of code for `dtypemeta.c` and all traversal loop and get_loop follow a very similar pattern.

ngoldbaum replied:
Thanks for clarifying!

seberg commented on numpy/core/src/multiarray/dtypemeta.c:
```diff
@@ -698,6 +735,39 @@ object_common_dtype(
+        *optr++ = zero;
+    }
+    Py_DECREF(zero);
+    if (PyErr_Occurred()) {
+        return -1;
+    }
```
```suggestion
```
An error cannot occur and if it did we should have checked explicitly for it.  The only thing that may have a check is `zero = `.  (cannot fail on CPython, but not sure about PyPy)

ngoldbaum replied:
OK good point, FWIW this was taken from the old implementation of `_zerofill`, it probably wasn't necessary there either.

seberg commented on numpy/core/src/multiarray/dtypemeta.c:
```diff
@@ -698,6 +735,39 @@ object_common_dtype(
+    PyObject **optr = (PyObject **)data;
+    while (size--) {
+        Py_INCREF(zero);
+        *optr++ = zero;
```
In order to correctly support unaligned memory (for structured dtypes) at least.  We probably need to use a `memcpy` here unfortunately.  (and a `char *` pointer)

ngoldbaum replied:
Ah my first pass over `PyArray_FillObjectArray` made me think that this is what it was doing, but it's actually doing a `memcpy`. Thanks for this catch also.

seberg reviewed (commented):
Thanks, happy to go this route.

One more thing to note: Your object filling loop does not DECREF anything that may be inside the array before.

We can do this, it just means:
* We need to make a note of the choice that the loop cannot be called on memory that holds references.  (You cannot use this to zerofill a used/fully initialized array, you would have to dealloc/clear it first!)
* In theory we could choose to avoid the initial memory zeroing due to references always.  But if we want to keep that open we would need to ask the zero-fill function to guarantee an initialized state on failure (in other words: calling clear on the memory must be OK).   Or find another safe scheme.
   (Not sure it matters in practice, for large arrays, we will touch all memory twice here, but the memory churn could be improved by us via chunking, and most zeroed arrays are likely used more than twice anyway.)

As I said, I can help with the void loop in a follow-up, or we can work together on it here (it may not be super hard by looking at the clear loop.  We might even be able to practically re-use it).

ngoldbaum committed:
MNT: respond to review comments

ngoldbaum committed:
MAINT: support traversal loops without initializing the descriptor

ngoldbaum committed:
MAINT: housecleaning for traversal loop setup and implementations

ngoldbaum committed:
DOC: warn against using zero-filling loop with a previously allocated array

ngoldbaum commented:
> We need to make a note of the choice that the loop cannot be called on memory that holds references.

Added a comment in the loop itself and added a warning in the documentation for the API hook in `dtypemeta.h`.

> I can help with the void loop in a follow-up

I'd prefer this, if that's alright. This is already 3 or 4 levels of yak shaving from where I started working last week :)

ngoldbaum commented:
Also I should highlight https://github.com/numpy/numpy/pull/23591/commits/4403c01fc2a60bf7d08a356c0c37e3a98a3df26a which made some changes to the setup for the traversal info struct. For zeroing we don't need the descriptor at all (the array creation routine already has it), so it's OK to just leave it as `NULL`.

seberg commented:
OK, thanks.  Lets put this in and I will fix the void loop in a follow-up.

(If anyone prefers another API, e.g. to ask for a single value only.  Happy to entertain that also.)
